File: OrdenamientoBurbuja.py
import time


def bubble_sort(arr):
    n = len(arr)
    for i in range(n - 1):
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]


def leer_archivo(nombre_archivo, n):
    try:
        with open(nombre_archivo, 'r') as archivo:
            todos_los_numeros = [int(num) for num in archivo.read().split()]
            return todos_los_numeros[:n]
    except FileNotFoundError:
        print(f"Error: El archivo '{nombre_archivo}' no se encontró.")
        return None
    except ValueError:
        print("Error: El archivo contiene datos que no son números enteros.")
        return None


def main():
    nombre_archivo = "datos.txt"

    try:
        n_str = input("Ingrese la cantidad de números a leer del archivo: ")
        n = int(n_str)
    except ValueError:
        print("Error: Ingrese un número entero válido.")
        return

    arreglo = leer_archivo(nombre_archivo, n)

    if arreglo is None:
        return

    print("\nDatos leídos del archivo:")
    print(*arreglo)

    start_time = time.perf_counter()
    bubble_sort(arreglo)
    end_time = time.perf_counter()

    tiempo = end_time - start_time

    print("\nDatos ordenados con Bubble Sort:")
    print(*arreglo)

    print(f"\nTiempo de ejecución Bubble Sort: {tiempo:.6f} segundos")


if __name__ == "__main__":
    main()

File: OrdenamientoBurbuja.c
// Bubble Sort

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void bubble_sort(int A[], int n) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (A[j] > A[j + 1]) {
                int aux = A[j];
                A[j] = A[j + 1];
                A[j + 1] = aux;
            }
        }
    }
}

int* leerArchivo(const char* nombreArchivo, int n, int* tamanoReal) {
    FILE* archivo = fopen(nombreArchivo, "r");
    if (!archivo) {
        printf("Error al abrir el archivo.\n");
        return NULL;
    }

    int* arreglo = (int*) malloc(n * sizeof(int));
    if (!arreglo) {
        fclose(archivo);
        printf("Error al asignar memoria.\n");
        return NULL;
    }

    int num, count = 0;
    while (fscanf(archivo, "%d", &num) == 1 && count < n) {
        arreglo[count++] = num;
    }

    fclose(archivo);
    *tamanoReal = count;
    return arreglo;
}

int main() {
    char nombreArchivo[] = "datos.txt";
    int n;

    printf("Ingrese la cantidad de números a leer del archivo: ");
    scanf("%d", &n);

    int tamanoReal;
    int* arreglo = leerArchivo(nombreArchivo, n, &tamanoReal);
    if (!arreglo) {
        return 1;
    }

    printf("\nDatos leídos del archivo:\n");
    for (int i = 0; i < tamanoReal; i++) {
        printf("%d ", arreglo[i]);
    }
    printf("\n");

    clock_t start = clock();
    bubble_sort(arreglo, tamanoReal);
    clock_t end = clock();

    double tiempo = (double)(end - start) / CLOCKS_PER_SEC;

    printf("\nDatos ordenados con Bubble Sort:\n");
    for (int i = 0; i < tamanoReal; i++) {
        printf("%d ", arreglo[i]);
    }
    printf("\n");

    printf("\nTiempo de ejecución Bubble Sort: %f segundos\n", tiempo);

    free(arreglo);
    return 0;
}
